### PR TITLE
Bugwork: microwave(general spawn) fix + client equipment fix

### LINKED ISSUE #790: Microwave only working every 1 out of 2 times
# Preliminary information

## Category
BUG

## Bug: Release or Development?
Test server 0.2.2A2

# Report
## Current Behavior
When using the microwave, the first try gives steak, the second does not, the third does and so on.

## Expected/Wanted/Requested Behavior
Meat in == Steak out.

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
1. Put meat in microwave
2. See output
3. Try again

## OS used:
OSX

## System specifications
macbook pro retina 2012 (late)

## Network
10-20ms ping

### PR DESCRIPTION
**Pull request by fomalsd**
### Purpose
Fixes #790 and #837

### Approach
#790 fix: CNT wasn't being initialized correctly on freshly netSpawned gameObjects; made it initialize properly
#837 fix: it was ServerMessage update that broke stuff indeed; introduced a coroutine that waits until player control is captured

### Open Questions and Pre-Merge TODOs

- [X]  I read the contribution guidelines and the wiki.
- [X]  This fix is tested on the same branch it is PR'ed to.
- [X]  I correctly commented my code
- [X]  This PR does not include files without specific need to do so.
- [X]  This PR does not bring up any new compile errors
- [X]  This PR has been tested in editor
- [X]  This PR has been tested in multiplayer

### Notes:
Will squash some more soon

### CODE REVIEW

**Comment by DooblyNoobly** [comment]
First test using two clients produced this on connection of client (clientside errors):
```
NullReferenceException: Object reference not set to an instance of an object
  at ForceJobListUpdateMessage+<Process>c__Iterator0.MoveNext () [0x00078] in /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/Messages/Server/ForceJobListUpdateMessage.cs:22 
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00028] in /Users/builduser/buildslave/unity/build/Runtime/Export/Coroutines.cs:17 
```
 
`(Filename: /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/Messages/Server/ForceJobListUpdateMessage.cs Line: 22)`

```
NullReferenceException: Object reference not set to an instance of an object
  at UpdateConnectedPlayersMessage.<ToString>m__0 (UnityEngine.GameObject p) [0x00001] in /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/Messages/Server/UpdateConnectedPlayersMessage.cs:60 
  at System.Linq.Enumerable+SelectArrayIterator`2[TSource,TResult].MoveNext () [0x0003a] in <11c1c8f1ba1046b191f565f88728429e>:0 
  at System.String.Join (System.String separator, System.Collections.Generic.IEnumerable`1[T] values) [0x0001f] in <9c9f068c46c64ffd91fda7af157b4d15>:0 
  at UpdateConnectedPlayersMessage.ToString () [0x00048] in /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/Messages/Server/UpdateConnectedPlayersMessage.cs:60 
  at System.Text.StringBuilder.AppendFormatHelper (System.IFormatProvider provider, System.String format, System.ParamsArray args) [0x002b1] in <9c9f068c46c64ffd91fda7af157b4d15>:0 
  at System.String.FormatHelper (System.IFormatProvider provider, System.String format, System.ParamsArray args) [0x00023] in <9c9f068c46c64ffd91fda7af157b4d15>:0 
  at System.String.Format (System.String format, System.Object arg0) [0x00008] in <9c9f068c46c64ffd91fda7af157b4d15>:0 
  at GameMessageBase+<WaitFor>c__Iterator0.MoveNext () [0x0003e] in /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/Messages/GameMessageBase.cs:17 
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00028] in /Users/builduser/buildslave/unity/build/Runtime/Export/Coroutines.cs:17 
UnityEngine.MonoBehaviour:StartCoroutine_Auto_Internal(MonoBehaviour, IEnumerator)
UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator) (at /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/MonoBehaviourBindings.gen.cs:62)
<RegisterHandler>c__AnonStorey0`1:<>m__0(NetworkMessage) (at /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/NetworkManagerExtensions.cs:46)
UnityEngine.Networking.NetworkConnection:HandleReader(NetworkReader, Int32, Int32) (at /Users/builduser/buildslave/unity/build/Extensions/Networking/Runtime/NetworkConnection.cs:469)
UnityEngine.Networking.NetworkConnection:HandleBytes(Byte[], Int32, Int32) (at /Users/builduser/buildslave/unity/build/Extensions/Networking/Runtime/NetworkConnection.cs:425)
UnityEngine.Networking.NetworkConnection:TransportReceive(Byte[], Int32, Int32) (at /Users/builduser/buildslave/unity/build/Extensions/Networking/Runtime/NetworkConnection.cs:576)
UnityEngine.Networking.NetworkClient:Update() (at /Users/builduser/buildslave/unity/build/Extensions/Networking/Runtime/NetworkClient.cs:753)
UnityEngine.Networking.NetworkClient:UpdateClients() (at /Users/builduser/buildslave/unity/build/Extensions/Networking/Runtime/NetworkClient.cs:965)
UnityEngine.Networking.NetworkIdentity:UNetStaticUpdate() (at /Users/builduser/buildslave/unity/build/Extensions/Networking/Runtime/NetworkIdentity.cs:1091)
```
 
`(Filename: /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/Messages/Server/UpdateConnectedPlayersMessage.cs Line: 60)`

**Comment by DooblyNoobly** [comment]
Also this happens in playerlog: 
image

**Comment by DooblyNoobly** [comment]
Then I killed the client, and reconnected still get NRE:

```
NullReferenceException: Object reference not set to an instance of an object
  at UpdateConnectedPlayersMessage.<ToString>m__0 (UnityEngine.GameObject p) [0x00001] in /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/Messages/Server/UpdateConnectedPlayersMessage.cs:60 
  at System.Linq.Enumerable+SelectArrayIterator`2[TSource,TResult].MoveNext () [0x0003a] in <11c1c8f1ba1046b191f565f88728429e>:0 
  at System.String.Join (System.String separator, System.Collections.Generic.IEnumerable`1[T] values) [0x0006b] in <9c9f068c46c64ffd91fda7af157b4d15>:0 
  at UpdateConnectedPlayersMessage.ToString () [0x00048] in /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/Messages/Server/UpdateConnectedPlayersMessage.cs:60 
  at System.Text.StringBuilder.AppendFormatHelper (System.IFormatProvider provider, System.String format, System.ParamsArray args) [0x002b1] in <9c9f068c46c64ffd91fda7af157b4d15>:0 
  at System.String.FormatHelper (System.IFormatProvider provider, System.String format, System.ParamsArray args) [0x00023] in <9c9f068c46c64ffd91fda7af157b4d15>:0 
  at System.String.Format (System.String format, System.Object arg0) [0x00008] in <9c9f068c46c64ffd91fda7af157b4d15>:0 
  at GameMessageBase+<WaitFor>c__Iterator0.MoveNext () [0x0003e] in /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/Messages/GameMessageBase.cs:17 
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00028] in /Users/builduser/buildslave/unity/build/Runtime/Export/Coroutines.cs:17 
UnityEngine.MonoBehaviour:StartCoroutine_Auto_Internal(MonoBehaviour, IEnumerator)
UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator) (at /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/MonoBehaviourBindings.gen.cs:62)
<RegisterHandler>c__AnonStorey0`1:<>m__0(NetworkMessage) (at /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/NetworkManagerExtensions.cs:46)
UnityEngine.Networking.NetworkConnection:InvokeHandler(NetworkMessage) (at /Users/builduser/buildslave/unity/build/Extensions/Networking/Runtime/NetworkConnection.cs:231)
UnityEngine.Networking.LocalClient:ProcessInternalMessages() (at /Users/builduser/buildslave/unity/build/Extensions/Networking/Runtime/LocalClient.cs:141)
UnityEngine.Networking.LocalClient:Update() (at /Users/builduser/buildslave/unity/build/Extensions/Networking/Runtime/LocalClient.cs:69)
UnityEngine.Networking.NetworkClient:UpdateClients() (at /Users/builduser/buildslave/unity/build/Extensions/Networking/Runtime/NetworkClient.cs:965)
UnityEngine.Networking.NetworkIdentity:UNetStaticUpdate() (at /Users/builduser/buildslave/unity/build/Extensions/Networking/Runtime/NetworkIdentity.cs:1091)
 
(Filename: /Users/logos/Desktop/unitystation/UnityProject/Assets/Scripts/Messages/Server/UpdateConnectedPlayersMessage.cs Line: 60)
```

File: UnityProject/Assets/Scripts/Messages/Server/ServerMessage.cs [hunk 1]
```diff
@@ -17,14 +17,13 @@ public void SendTo(GameObject recipient)
 	{
 		NetworkIdentity netIdentity = recipient.GetComponent<NetworkIdentity>();
 
-		//Only send to the client of the currently owned player as some
-		//netID's being used in this method could be dead players, exclude them:
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/Messages/Server/ServerMessage.cs]: The comments I put here were trying to explain that the old way was sending the message to the connection of each dead person as well. Now we just send it to the connection of a player that is in the playerlist instead which makes sure we are only sending the message once to the client. I felt people would lose the meaning of why I made that condition over time

**Review by DooblyNoobly — DISMISSED** [review]
- Investigate why messages on start broke
- Remove the massive NUL spam in player.log

**Comment by fomalsd** [comment]
No errors happen for me, even with 2 clients, perhaps recheck or someone else try this PR?

**Comment by fomalsd** [comment]
didn't have a chance to look into this yet, will do today/tomorrow

**Comment by fomalsd** [comment]
No kidding, couldn't reproduce any of these.
Tried following environments:
-headed host(editor) + two 200ms clients (builds), all from OutpostDM (no round restarts)
-headed host(editor) + two 200ms clients (builds), all from Lobby
-"test-server headless" host(editor) + two 200ms clients (builds), all from Lobby
no errors on respawn:
default

Everything tested on Windows.

The only errors I've got were:
1) upon joining: SyncList not initialized error, once for each client's player
`SyncList not initialized
UnityEngine.DebugLogHandler:Internal_Log(LogType, String, Object)
UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:LogError(Object)
UnityEngine.Networking.SyncList1:SendMsg(Operation, Int32, Int32) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\SyncList.cs:295)
UnityEngine.Networking.SyncList1:SendMsg(Operation, Int32) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\SyncList.cs:338)
UnityEngine.Networking.SyncList1:Clear() (at C:\buildslave\unity\build\Extensions\Networking\Runtime\SyncList.cs:395)
UnityEngine.Networking.SyncListInt:ReadReference(NetworkReader, SyncListInt) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\SyncList.cs:124)
IDCard:OnDeserialize(NetworkReader, Boolean)
UnityEngine.Networking.NetworkIdentity:OnUpdateVars(NetworkReader, Boolean) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkIdentity.cs:750)
UnityEngine.Networking.ClientScene:ApplySpawnPayload(NetworkIdentity, Vector3, Byte[], NetworkInstanceId, GameObject) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\ClientScene.cs:491)
UnityEngine.Networking.ClientScene:OnObjectSpawnScene(NetworkMessage) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\ClientScene.cs:608)
UnityEngine.Networking.NetworkConnection:HandleReader(NetworkReader, Int32, Int32) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkConnection.cs:469)
UnityEngine.Networking.NetworkConnection:HandleBytes(Byte[], Int32, Int32) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkConnection.cs:425)
UnityEngine.Networking.NetworkConnection:TransportReceive(Byte[], Int32, Int32) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkConnection.cs:576)
UnityEngine.Networking.NetworkClient:Update() (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkClient.cs:753)
UnityEngine.Networking.NetworkClient:UpdateClients() (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkClient.cs:965)
UnityEngine.Networking.NetworkIdentity:UNetStaticUpdate() (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkIdentity.cs:1091)
(Filename: C:/buildslave/unity/build/Extensions/Networking/Runtime/SyncList.cs Line: 295)`

2) on round restart: flavours of these (for each object, like >300 per player)
`Coroutine couldn't be started because the the game object 'Closet_Locker_Medical_Blue' is inactive!
UnityEngine.MonoBehaviour:StartCoroutine_Auto_Internal(MonoBehaviour, IEnumerator)
UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator) (at C:\buildslave\unity\build\artifacts\generated\common\runtime\MonoBehaviourBindings.gen.cs:62)
PushPull:OnStartClient() (at C:\homeproj\unitystation\UnityProject\Assets\Scripts\Objects\PushPull.cs:29)
UnityEngine.Networking.NetworkIdentity:OnStartClient() (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkIdentity.cs:374)
UnityEngine.Networking.ClientScene:ApplySpawnPayload(NetworkIdentity, Vector3, Byte[], NetworkInstanceId, GameObject) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\ClientScene.cs:505)
UnityEngine.Networking.ClientScene:OnObjectSpawnScene(NetworkMessage) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\ClientScene.cs:608)
UnityEngine.Networking.NetworkConnection:HandleReader(NetworkReader, Int32, Int32) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkConnection.cs:469)
UnityEngine.Networking.NetworkConnection:HandleBytes(Byte[], Int32, Int32) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkConnection.cs:425)
UnityEngine.Networking.NetworkConnection:TransportReceive(Byte[], Int32, Int32) (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkConnection.cs:576)
UnityEngine.Networking.NetworkClient:Update() (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkClient.cs:753)
UnityEngine.Networking.NetworkClient:UpdateClients() (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkClient.cs:965)
UnityEngine.Networking.NetworkIdentity:UNetStaticUpdate() (at C:\buildslave\unity\build\Extensions\Networking\Runtime\NetworkIdentity.cs:1091)
[C:\buildslave\unity\build\Runtime/Mono/MonoBehaviour.cpp line 758] 
(Filename: C:/buildslave/unity/build/artifacts/generated/common/runtime/MonoBehaviourBindings.gen.cs Line: 62)`

3) Occasional job list NREs for server
`NullReferenceException: Object reference not set to an instance of an object
ForceJobListUpdateMessage+<Process>c__Iterator0.MoveNext () (at Assets/Scripts/Messages/Server/ForceJobListUpdateMessage.cs:22)
UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) (at C:/buildslave/unity/build/Runtime/Export/Coroutines.cs:17)`

All of these exist on latest develop!

**Comment by fomalsd** [comment]
ahh, finally caught respawn bug: it's caused by bogus renaming algo (when two players join with same name). It most likely causes #847, too
will fix

**Comment by fomalsd** [comment]
making a playerlist rewrite, hope it will be ready by tomorrow if all goes well

**Comment by fomalsd** [comment]
will really try to finish this as soon as I can, still got a few small bugs to squash (as PlayerList had its guts replaced)

**Comment by fomalsd** [comment]
Ready to be reviewed!
Tested w/pseudo-headless and 3 clients (200ms)

**Pull request merged**